pkg/dns/util/util.go

```diff
@@ -41,15 +43,64 @@ const (
 
 // ExtractIP turns a standard PTR reverse record lookup name
 // into an IP address
-func ExtractIP(reverseName string) (string, bool) {
-	if !strings.HasSuffix(reverseName, ArpaSuffix) {
-		return "", false
+// Returns "", error if the reverseName is not a valid PTR lookup name
+func ExtractIP(reverseName string) (string, error) {
+	if strings.HasSuffix(reverseName, ArpaSuffix) {
+		ip, err := extractIPv4(strings.TrimSuffix(reverseName, ArpaSuffix))
+		if err != nil {
+			return "", fmt.Errorf("incorrect PTR IPv4 %q: %w", reverseName, err)
+		}
+		return ip, nil
 	}
-	search := strings.TrimSuffix(reverseName, ArpaSuffix)
 
+	if strings.HasSuffix(reverseName, ArpaSuffixV6) {
+		ip, err := extractIPv6(strings.TrimSuffix(reverseName, ArpaSuffixV6))
+		if err != nil {
+			return "", fmt.Errorf("incorrect PTR IPv6 %q: %w", reverseName, err)
+		}
+		return ip, nil
+	}
+
+	return "", fmt.Errorf("incorrect PTR: %q", reverseName)
+}
+
+// extractIPv4 turns a standard PTR reverse record lookup name
+// into an IP address
+func extractIPv4(reverseName string) (string, error) {
 	// reverse the segments and then combine them
-	segments := ReverseArray(strings.Split(search, "."))
-	return strings.Join(segments, "."), true
+	segments := ReverseArray(strings.Split(reverseName, "."))
+
+	ip := net.ParseIP(strings.Join(segments, ".")).To4()
+	if ip == nil {
+		return "", fmt.Errorf("failed to parse IPv4 reverse name: %q", reverseName)
+	}
+	return ip.String(), nil
+}
+
+// extractIPv6 turns a IPv6 PTR reverse record lookup name
+// into an IPv6 address according to RFC3596
+// b.a.9.8.7.6.5.0.4.0.0.0.3.0.0.0.2.0.0.0.1.0.0.0.0.0.0.0.1.2.3.4.ip6.arpa.
+// is reversed to 4321:0:1:2:3:4:567:89ab
+func extractIPv6(reverseName string) (string, error) {
+	segments := ReverseArray(strings.Split(reverseName, "."))
+
+	// IPv6nibbleCount is the expected number of nibbles in IPv6 PTR record as defined in rfc3596
+	const ipv6nibbleCount = 32
+
+	if len(segments) != ipv6nibbleCount {
+		return "", fmt.Errorf("incorrect number of segments in IPv6 PTR: %v", len(segments))
+	}
+
+	var slice6 []string
+	for i := 0; i < len(segments); i += 4 {
```

Comment:
nit: you can use `ipv6nibbleCount`

Reply:
We iterate over segments slice so in the loop condition it's more clean to use `len(segments)` as well IMO